core/src/main/java/com/datastrato/gravitino/catalog/CatalogOperationsProxy.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2024 Datastrato Pvt Ltd.
+ * This software is licensed under the Apache License version 2.
+ */
+package com.datastrato.gravitino.catalog;
+
+import com.datastrato.gravitino.utils.PrincipalUtils;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.Collections;
+
+/** Proxy wrapper on CatalogOperations to execute operations by impersonating given user */
+public class CatalogOperationsProxy implements InvocationHandler {
+
+  private final CatalogProxyPlugin plugin;
+  private final CatalogOperations ops;
+
+  private CatalogOperationsProxy(CatalogProxyPlugin plugin, CatalogOperations ops) {
+    this.plugin = plugin;
+    this.ops = ops;
+  }
+
+  public static <T extends CatalogOperations> T getProxy(T ops, CatalogProxyPlugin plugin) {
```

Comment:
Also please check @mchades 's partition API design, it is implemented in table level, not in catalog level, so this way may not be enough.

Reply:
Maybe we can add a partition proxy for that. I can create a new issue to track it.

Reply:
Renamed this class to OperationsProxy to adapt more scenaries.